syntax = "proto3";

package redpepper;

enum MessageType {
    UNSPECIFIED = 0;
    CLIENTHELLO = 10;
    SERVERHELLO = 11;
    BYE = 12;
    PING = 13;
    PONG = 14;

    COMMAND = 20;
    COMMANDCANCEL = 21;
    COMMANDPROGRESS = 22;
    COMMANDRESULT = 23;

    REQUEST = 30;
    RESPONSE = 31;
}

message Message {
    MessageType type = 1;
    oneof message {
        ClientHello client_hello = 10;
        ServerHello server_hello = 11;
        Bye bye = 13;
        Ping ping = 14;
        Pong pong = 15;

        Command command = 20;
        CommandCancel cancel = 21;
        CommandProgress progress = 22;
        CommandResult result = 23;

        Request request = 30;
        Response response = 31;
    }
}

message Bye {
    string reason = 1;
}

message ClientHello {
    string clientID = 1;
    string auth = 2;
}

message ServerHello {
    uint32 version = 1;
}

message Ping {
    int32 data = 1;
}

message Pong {
    int32 data = 1;
}

message Command {
    int64 commandID = 1;
    string type = 2;
    string data = 3;
}

message CommandCancel {
    int64 commandID = 1;
}

message CommandResult {
    int64 commandID = 1;
    enum Status {
        UNSPECIFIED = 0;
        SUCCESS = 1;
        FAILED = 2;
        CANCELLED = 3;
    }
    Status status = 2;
    bool changed = 3;
    string output = 5;
}

message CommandProgress {
    int64 commandID = 1;
    uint32 current = 2;
    uint32 total = 3;
    string message = 4;
}

message Request {
    int64 requestID = 1;
    string name = 2;
    string data = 3;
}

message Response {
    int64 requestID = 1;
    bool success = 2;
    string data = 3;
}

